### PR TITLE
python312Packages.tree-sitter_0_21: remove from propagated dependencies

### PR DESCRIPTION
**Pull request by natsukium**
## Description of changes

Python packages must not propagate such versioned packages due to PYTHONPATH conflicts.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/tree-sitter-languages/default.nix [hunk 1]
```diff
@@ -37,17 +38,12 @@ buildPythonPackage rec {
     wheel
```
Comment (dotlambda) [inline pkgs/development/python-modules/tree-sitter-languages/default.nix:38]:
```suggestion
```

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 332747` at a930fa80 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.grep-ast</li>
    <li>python311Packages.tree-sitter-languages</li>
    <li>python312Packages.grep-ast</li>
    <li>python312Packages.tree-sitter-languages</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>browsr</li>
    <li>textual-paint</li>
  </ul>
</details>
<details>
  <summary>27 packages built successfully:</summary>
  <ul>
    <li>autotools-language-server</li>
    <li>dooit</li>
    <li>frogmouth</li>
    <li>memray</li>
    <li>oterm</li>
    <li>python311Packages.graphrag</li>
    <li>python311Packages.lsp-tree-sitter</li>
    <li>python311Packages.manifestoo</li>
    <li>python311Packages.pytest-textual-snapshot</li>
    <li>python311Packages.textual</li>
    <li>python311Packages.textual-dev</li>
    <li>python311Packages.textual-universal-directorytree</li>
    <li>python311Packages.tree-sitter_0_21</li>
    <li>python312Packages.graphrag</li>
    <li>python312Packages.lsp-tree-sitter</li>
    <li>python312Packages.manifestoo</li>
    <li>python312Packages.pytest-textual-snapshot</li>
    <li>python312Packages.textual</li>
    <li>python312Packages.textual-dev</li>
    <li>python312Packages.textual-universal-directorytree</li>
    <li>python312Packages.tree-sitter_0_21</li>
    <li>rich-cli</li>
    <li>smassh</li>
    <li>tftui</li>
    <li>toolong</li>
    <li>upiano</li>
    <li>wsrepl</li>
  </ul>
</details>

**Review by doronbehar — COMMENTED** [review]
Changes look good overall, I wonder how come so many packages got dependent on `tree-sitter_0_21`... I'm only a bit confused about the changes to `python31{1,2}Packages.textual`:

The `optional-dependencies.syntax` attribute is confusing to me - why not change there the `tree-sitter` dependency to `tree-sitter_0_21`? If the last commit will be included eventually, the `optional-dependencies.syntax` attribute will almost never be evaluated, but it is still confusing, especially since this PR is meant to avoid mixing different versions of the same python package in the dependency tree of packages.

Another thing that bothers me, is the fact that without the last commit, `python312Packages.textual` builds fine on Hydra. However, `python311Packages.textual` doesn't evaluate because 
`tree-sitter-languages` is marked as broken on Darwin. Although the `lib.optionals` condition should not add it to `optional-dependencies.syntax`. Do you know how to explain that please @natsukium ?

Also, if the goal of the last commit is to fix the <code>python31<b>1</b>Packages.textual</code>, the commit message should indicate that.

**Comment by doronbehar** [comment]
> Changes look good overall, I wonder how come so many packages got dependent on `tree-sitter_0_21`...

I take these words back... I iterated once more the commits and I do understand why these packages dependent on `tree-sitter_0_21` - the comments explained this.

I would approve a PR that would change only `python31{1,2}Packages.textual`, because that is the only package that has mixed `tree-sitter` versions in it's dependencies.

**Comment by natsukium** [comment]
Did you notice that `grep-ast` depends on `tree-sitter_0_21` through `tree-sitter-languages` in https://github.com/NixOS/nixpkgs/pull/331173 ?
Also, note that `tree-sitter_0_21` will be propagated when a certain package adds `textual.optional-dependencies.syntax` to its dependencies.

We can't control the downstream packages or the user environment.
So the current `tree-sitter-languages` and `lsp-tree-sitter` are unacceptable.

**Comment by natsukium** [comment]
I found that `textual` does not require `tree-sitter-languages` for testing and removed it from nativeCheckInputs.

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 332747` at 3efc201b run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.grep-ast</li>
    <li>python311Packages.tree-sitter-languages</li>
    <li>python312Packages.grep-ast</li>
    <li>python312Packages.tree-sitter-languages</li>
  </ul>
</details>
<details>
  <summary>15 packages failed to build:</summary>
  <ul>
    <li>browsr</li>
    <li>frogmouth</li>
    <li>oterm</li>
    <li>python312Packages.graphrag</li>
    <li>python312Packages.manifestoo</li>
    <li>python312Packages.pytest-textual-snapshot</li>
    <li>python312Packages.textual</li>
    <li>python312Packages.textual-dev</li>
    <li>python312Packages.textual-universal-directorytree</li>
    <li>python312Packages.tree-sitter_0_21</li>
    <li>rich-cli</li>
    <li>textual-paint</li>
    <li>tftui</li>
    <li>upiano</li>
    <li>wsrepl</li>
  </ul>
</details>
<details>
  <summary>13 packages built successfully:</summary>
  <ul>
    <li>autotools-language-server</li>
    <li>dooit</li>
    <li>python311Packages.graphrag</li>
    <li>python311Packages.lsp-tree-sitter</li>
    <li>python311Packages.manifestoo</li>
    <li>python311Packages.pytest-textual-snapshot</li>
    <li>python311Packages.textual</li>
    <li>python311Packages.textual-dev</li>
    <li>python311Packages.textual-universal-directorytree</li>
    <li>python311Packages.tree-sitter_0_21</li>
    <li>python312Packages.lsp-tree-sitter</li>
    <li>smassh</li>
    <li>toolong</li>
  </ul>
</details>

**Comment by doronbehar** [comment]
> Also, note that `tree-sitter_0_21` will be propagated when a certain package adds `textual.optional-dependencies.syntax` to its dependencies.

Yea indeed it is a weird thing to do.

> Did you notice that `grep-ast` depends on `tree-sitter_0_21` through `tree-sitter-languages` in #331173 ?

No.. (perhaps I should have added my self to the maintainers.)

> So the current `tree-sitter-languages` and `lsp-tree-sitter` are unacceptable.

OK, that is convincing.

> I found that `textual` does not require `tree-sitter-languages` for testing and removed it from nativeCheckInputs.

That's good. I took the liberty to rebase the PR onto latest `master`, so it is a bit more clear what is the state of things before and after and changed a bit the commit order, and the `textual` commit's message.

**Comment by natsukium** [comment]
Thanks for understanding and adding context.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -15814,7 +15814,7 @@ self: super: with self; {
 
   tree-sitter = callPackage ../development/python-modules/tree-sitter { };
 
-  tree-sitter0_21 = callPackage ../development/python-modules/tree-sitter0_21 { };
+  tree-sitter_0_21 = callPackage ../development/python-modules/tree-sitter/0_21.nix { };
```
Comment (eclairevoyant) [inline pkgs/top-level/python-packages.nix:15817]: Would have been ideal to add an entry in python-aliases for the rename.

Comment (natsukium) [reply pkgs/top-level/python-packages.nix:15817]: Certainly, but it's an internal package and I didn't want to pollute the namespace too much.

Comment (doronbehar) [reply pkgs/top-level/python-packages.nix:15817]: > > Would have been ideal to add an entry in python-aliases for the rename.
> Certainly, but it's an internal package and I didn't want to pollute the namespace too much.

Also, that attribute was there for a ~month or less, and it is an outdated version, so I barely think anyone had started using it externally.

**Pull request merged**